packages/core/src/odata/common/request-builder/batch/batch-response-transformer.ts

```diff
@@ -0,0 +1,220 @@
+      ];
+    }
+
+    logger.warn('Could not parse constructor from response body.');
+  }
+}
+
+/**
+ * Transform the parsed batch response.
+ * @param parsedBatchResponse Two dimensional list of parsed batch sub responses.
+ * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns An array of parsed sub responses of the batch response.
+ */
+export function transformBatchResponse(
+  parsedBatchResponse: (ResponseData[] | ResponseData)[],
+  entityToConstructorMap: Record<string, Constructable<EntityBase>>,
+  responseDataAccessor: ResponseDataAccessor,
+  deserializer: EntityDeserializer
+): (ErrorResponse | ReadResponse | WriteResponses)[] {
+  const transformer = new BatchResponseTransformer(
+    entityToConstructorMap,
+    responseDataAccessor,
+    deserializer
+  );
+  return transformer.transformBatchResponse(parsedBatchResponse);
+}
+
+/**
+ * Create a function to transform the parsed response body to a list of entities of the given type or an error.
+ * @param body The parsed JSON reponse body.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns A function to be used for transformation of the read response.
+ */
+function asReadResponse(
+  body: any,
+  responseDataAccessor: ResponseDataAccessor,
+  deserializer: EntityDeserializer
+) {
+  return <EntityT extends EntityBase>(
+    constructor: Constructable<EntityT>
+  ): Error | EntityT[] => {
+    if (body.error) {
+      return new Error(body.error);
+    }
+    if (responseDataAccessor.isCollectionResult(body)) {
+      return responseDataAccessor
+        .getCollectionResult(body)
+        .map(r => deserializer.deserializeEntity(r, constructor));
+    }
+    return [
+      deserializer.deserializeEntity(
+        responseDataAccessor.getSingleResult(body),
+        constructor
+      )
+    ];
+  };
+}
+
+/**
+ * Create a function to transform the parsed response body to an entity of the given type.
+ * @param body The parsed JSON reponse body.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns A function to be used for transformation of the write response.
+ */
+function asWriteResponse(
+  body: any,
+  responseDataAccessor: ResponseDataAccessor,
+  deserializer: EntityDeserializer
+) {
+  return <EntityT extends EntityBase>(constructor: Constructable<EntityT>) =>
+    deserializer.deserializeEntity(
+      responseDataAccessor.getSingleResult(body),
+      constructor
+    );
+}
+
+/**
+ * Parse the entity name from the metadata uri. This should be the `__metadata` property of a single entity in the response.
+ * @param uri The URI to parse the entity name from
+ * @returns The entity name.
+ */
+export function parseEntityNameFromMetadataUri(
+  uri: string | undefined
```

Comment:
in case of `undefined` an exception is thrown. For am exported method I think this is not a good signature. If it say `string | undefined` it should no throw something if I pass undefined.

Reply:
I would see this independent of whether this is exported or not, but you're right, this is overkill. I added this because I thought if the input is object.key and the key is not defined it might be helpful, but it compiles independently of this.